### DISCUSSION TITLE
How to override AdminContainer.razor?

### DISCUSSION THREAD

**Discussion by lthoa** [oqtane/oqtane.framework#1055] [Q&A]
I created the AdminContainer.razor file in my theme project, but it still uses default AdminContainer from the Oqtane theme.

Thanks!

**Comment by leigh-pointer** [comment]
Next to the login button you should see a pencil icon. click this
The module will then change in to admin mode.   On the top left is a small down arrow, click that to activate the context menu.
Select "Manage settings" on the module settings is the option to select what container you wish for a module.

**Reply by lthoa** [reply]
Thanks @leigh-pointer. That is selection for module's content view. 
I want to change the view of Edit page only.

**Reply by leigh-pointer** [reply]
Currently I don't think that functionality exists.  I have checked the site setting's where the default theme is set, but no option for Administration.

**Reply by sbwalker** [reply]
@lthoa @leigh-pointer is correct - this functionality does not currently exist. There is a feature in the IModuleControl interface where you can specify UseAdminContainer = false ( it is true by default ) which tells Oqtane to use the container that is specified in the UI rather than the default AdminContainer.razor. However this solution would require you to change module code in order to affect the behavior. It sounds like what you were expecting is a convention where a theme can include their own AdminContainer.razor and the framework uses this instead of the default. I do like conventions but they tend to be non-intuitive because they are not explicitly defined anywhere ( ie. they are "magic" ). Another option would be to add a property to the ITheme interface for AdminContainer which allows the developer to optionally specify an alternative container for admin scenarios when that theme is in use. Similarly there could even be a DefaultContainer property that can optionally be specified so that the UI suggests a default non-admin container in the UI.

**Reply by PavelVsl** [reply]
Can be solved by set UseAdminContainer = false and write dialog code as you like in module.
for example my html editor:
```c#
@inherits HtmlTextBase

<ModalEditor>
    <Header>
        <h5 class="modal-title">
            <ModuleTitle/>
        </h5>
        <NavLink class="btn btn-secondary" href="@NavigateUrl()">&times;</NavLink>
    </Header>
    <Body>
    <div class="form-group h-100 p-0 m-0">
        <textarea class="text-monospace form-control h-100 text-nowrap" @bind="@Model" ></textarea>
    </div>

    </Body>
    <Footer>
        <button type="button" class="btn btn-success" @onclick="Save">Uložit</button>
        <NavLink class="btn btn-secondary" href="@NavigateUrl()">Storno</NavLink>

    </Footer>
</ModalEditor>


@code{
    public override bool UseAdminContainer => false;
    public override SecurityAccessLevel SecurityAccessLevel => SecurityAccessLevel.Admin;

}
```
and ModalEditor.razor source
```c#

<div class="mfl-admin-modal">
    <div class="modal" tabindex="-1" role="dialog">
        <div class="modal-dialog">
            <div class="modal-content">
                <div class="modal-header">
                        @Header
                </div>
                <div class="modal-body overflow-auto ">
                    @Body
                </div>
                <div class="modal-footer">
                    @Footer
                </div>
            </div>
        </div>
    </div>
</div>      

@code{
    [Parameter]
    public RenderFragment Header { get; set; }
        
    [Parameter]
    public RenderFragment Body { get; set; }
        
    [Parameter]
    public RenderFragment Footer { get; set; }
}

```

**Comment by sbwalker** [comment]
@chlupac I could be wrong but I think @lthoa wants to use a customized AdminContainer for every module in his installation - even the admin modules. And he does not want to modify the core code for every razor component in order to accomplish this. The theme enhancement I suggested would allow for this.

**Reply by PavelVsl** [reply]
we have discussion on this topic here #561

**Reply by sbwalker** [reply]
Yes, and at the time you said you solved your own problem using the existing property so we never pursued it any further. At that time I also did not understand what you meant by your comment "can be easily replaced with custom implementation from theme" in #561. Perhaps you were referring to something similar to what I described above?

**Reply by PavelVsl** [reply]
Yes,  we have to implement GetAdminContainerFromTheme()
```c#
        _moduleState = Module; // passed in from Pane component
        string container = _moduleState.ContainerType;
        if (PageState.ModuleId != -1 && PageState.Action != "" && _moduleState.UseAdminContainer)
        {
            //container = Constants.DefaultAdminContainer;
            container = GetAdminContainerFromTheme() ?? Constants.DefaultAdminContainer;
        }
```

**Reply by lthoa** [reply]
> @chlupac I could be wrong but I think @lthoa wants to use a customized AdminContainer for every module in his installation - even the admin modules. And he does not want to modify the core code for every razor component in order to accomplish this. The theme enhancement I suggested would allow for this.

You're right @sbwalker, I want to use a custom AdminContainer for all my modules , and hopefully it's supported by default in Oqtane. When designing a theme, I want it to have both a consistent UI and Admin. The ability to customize AdminContainer is essential for developers, and it's a pity not to provide it in the platform.
At the moment @chlupac 's solution to create ModalEditor control is the most plausible when I don't want to change the core code.

Thanks @sbwalker, @chlupac

**Comment by sbwalker** [comment]
@lthoa this has been resolved in #1088 - there is now the ability at the Site level to specify the Admin Container through the Site Settings UI.
